packages/@aws-cdk/aws-kinesisfirehose/lib/delivery-stream.ts

```diff
@@ -0,0 +1,256 @@
+   */
+  readonly deliveryStreamName?: string;
+
+
+  /**
+   * The IAM role associated with this delivery stream.
+   *
+   * Assumed by Kinesis Data Firehose to read from sources, invoke processors, and write to destinations.
+   *
+   * @default - the imported stream cannot be granted access to other resources as an `iam.IGrantable`.
+   */
+  readonly role?: iam.IRole;
+}
+
+/**
+ * Create a Kinesis Data Firehose delivery stream
+ *
+ * @resource AWS::KinesisFirehose::DeliveryStream
+ */
+export class DeliveryStream extends DeliveryStreamBase {
+  /**
+   * Import an existing delivery stream from its name.
+   */
+  static fromDeliveryStreamName(scope: Construct, id: string, deliveryStreamName: string): IDeliveryStream {
+    return this.fromDeliveryStreamAttributes(scope, id, { deliveryStreamName });
+  }
+
+  /**
+   * Import an existing delivery stream from its ARN.
+   */
+  static fromDeliveryStreamArn(scope: Construct, id: string, deliveryStreamArn: string): IDeliveryStream {
+    return this.fromDeliveryStreamAttributes(scope, id, { deliveryStreamArn });
+  }
+
+  /**
+   * Import an existing delivery stream from its attributes.
+   */
+  static fromDeliveryStreamAttributes(scope: Construct, id: string, attrs: DeliveryStreamAttributes): IDeliveryStream {
+    if (!attrs.deliveryStreamName && !attrs.deliveryStreamArn) {
+      throw new Error('Either deliveryStreamName or deliveryStreamArn must be provided in DeliveryStreamAttributes');
+    }
+    const deliveryStreamName = attrs.deliveryStreamName ?? cdk.Stack.of(scope).parseArn(attrs.deliveryStreamArn!).resourceName;
+    if (!deliveryStreamName) {
+      throw new Error(`Could not import delivery stream from malformatted ARN ${attrs.deliveryStreamArn}: could not determine resource name`);
+    }
+    const deliveryStreamArn = attrs.deliveryStreamArn ?? cdk.Stack.of(scope).formatArn({
+      service: 'firehose',
+      resource: 'deliverystream',
+      resourceName: attrs.deliveryStreamName,
+    });
+    class Import extends DeliveryStreamBase {
+      public readonly deliveryStreamName = deliveryStreamName!;
+      public readonly deliveryStreamArn = deliveryStreamArn;
+      public readonly grantPrincipal = attrs.role ?? new iam.UnknownPrincipal({ resource: this });
+    }
+    return new Import(scope, id);
+  }
+
+  readonly deliveryStreamName: string;
+
+  readonly deliveryStreamArn: string;
+
+  readonly grantPrincipal: iam.IPrincipal;
+
+  constructor(scope: Construct, id: string, props: DeliveryStreamProps) {
+    super(scope, id);
+
+    const role = props.role ?? new iam.Role(this, 'Service Role', {
+      assumedBy: new iam.ServicePrincipal('firehose.amazonaws.com'),
+    });
+    this.grantPrincipal = role;
+
+    if (props.destinations.length !== 1) {
+      throw new Error(`Only one destination is allowed per delivery stream, given ${props.destinations.length}`);
+    }
+    const destinationConfig = props.destinations[0].bind(this, { deliveryStream: this });
```

Comment:
I don't think this `bind()` is correctly modeled.

Notice that you pass `this` _before_ you assign properties like `this. deliveryStreamArn` and `this. deliveryStreamName`. Which means that if the destination accesses them, they will get `undefined`, even though their type is `string`, and will most likely crash.

There are 2 ways out of this:

1. Don't pass `DeliveryStream` to `bind()`, just `Construct` (many `bind()`s work like this).
2. Create the `CfnDeliveryStream` with a Lazy for the destination, [like this](https://github.com/aws/aws-cdk/blob/94f81c441f9e2cb8dc70eb2e772d2cd75e468b67/packages/%40aws-cdk/aws-codebuild/lib/project.ts#L1084-L1091), and only call `bind()` on the Destination after you've set all properties like `this. deliveryStreamArn` and `this. deliveryStreamName`.

Your call!

Reply:
For # 1: In this case, I don't think I can only pass `Construct` since there are fields on the `DeliveryStream` that are needed by the bind. Maybe I should pass those in as `DestinationBindOptions`?

For #2: will take stab at this approach.

Reply:
> For # 1: In this case, I don't think I can only pass `Construct` since there are fields on the `DeliveryStream` that are needed by the bind. Maybe I should pass those in as `DestinationBindOptions`?

Yes, that's exactly what I meant.

Reply:
I implemented #1. We will need to add a Connections option when we support Redshift, but it shouldn't be a problem to add that to the interface later